### PR TITLE
Enable building on non-x86 and less than SSE4.1

### PR DESCRIPTION
**Pull request by mr-c**
- [x] needs a git submodule of https://github.com/nemequ/simde/ or cmake magic to pull that down
- [x] isnan on clang 10 https://github.com/nemequ/simde/issues/282
- [x] ISO C++ does not allow C99 designated initializers https://github.com/nemequ/simde/issues/287
- [x] ISO C++ forbids compound-literals https://github.com/nemequ/simde/issues/288
- [x] macOS clang, AppleClang 11.0.0.11000033, [has a weird compile error](https://dev.azure.com/themartinsteinegger/edc8daa1-9e01-4171-b135-27e374460be3/_apis/build/builds/592/logs/23) Fixed by upgrading to macOS 11.05 AppleClang 11.0.3.11030032
- [x] GCC+ASan crashes with an [internal error somewhere completely random](https://dev.azure.com/themartinsteinegger/mmseqs2/_build/results?buildId=609&view=logs&j=ab24d926-1cfd-5268-5232-547307163af3&t=f30e022a-6aef-58fb-62a5-07e89b64b206&l=749) https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=961011
- [ ] arm64 performance

### CODE REVIEW

**Comment by milot-mirdita** [comment]
Thanks a lot! I would suggest to temporarily remove `-DENABLE_WERROR=1` from the cmake calls in Azure and Travis. I'll try to look at your changes in more detail later.

I tried the debian binaries on Friday. x64 worked fine, however the ARM build failed in the regression test. I ran it inside QEMU during the night. I hope Travis and the other changes I had done based on sse2neon will be helpful to solve the test failure.

**Comment by milot-mirdita** [comment]
Could you rebase ontop of 67d61013628911d8fce01a5aed6380f2a3de7f23? I think that should solve some of the weird log length issues with travis.

**Comment by mr-c** [comment]
@milot-mirdita here goes! :-)

**Comment by milot-mirdita** [comment]
The current issues seem to be:
* There seems to be an issue with older gcc and  the use of`isnan` in SIMDe
* Alignments on both clang and gcc get stuck eventually
* macOS clang has a weird compile error
* GCC+ASan crashes with an internal error somewhere completely random (might be solved by using a newer gcc?)

I don't have time to look at the issues until later this week though :/

**Comment by junaruga** [comment]
> The current issues seem to be:

I think if the issues are related to SIMDe on x86 CPU. We can avoid some of the issues by enabling SIMDe only on non-x86 CPU architectures like this for now.

https://github.com/lh3/minimap2/blob/master/ksw2_extd2_sse.c

```
#ifdef __SSE2__
#ifdef USE_SIMDE
#include <simde/x86/sse2.h>
#else
#include <emmintrin.h>
#endif
...
#ifdef __SSE4_1__
#ifdef USE_SIMDE
#include <simde/x86/sse4.1.h>
#else
#include <smmintrin.h>
#endif
#endif
```

**Comment by milot-mirdita** [comment]
I tried compiling your fork and the error/warning list seems quite daunting. Instead I took a look at what instructions are missing for SSE2 support and found we are using exactly three ones that are only available in SSS3/SSE4.1:
* _mm_shuffle_epi8
* _mm_minpos_epu16
* _mm_extract_epi8

Porting only these instructions back from SIMDe is super easy. For Debian we then would produce only three binaries: mmseqs-sse2 mmseqs-sse41 and mmseqs-avx2. Would that be okay?

How does Debian produce binaries for architectures other than x86(_64), ARM(64) and PPCLE? SIMDe seems to only support those (and ppcle only partially). https://packages.debian.org/sid/mmseqs2 shows a lot of architectures.

**Comment by milot-mirdita** [comment]
Ah sorry I didn't really pay attention, SIMDe does support more architectures.

**Comment by mr-c** [comment]
>  Instead I took a look at what instructions are missing for SSE2 support 

I think more those are required for not-X86 ?

I've force pushed to this branch [with a newer SIMDe version](https://github.com/nemequ/simde/compare/0a0cdd20c9ac4f0cd787f60cca30c0aa1e71b6c1...aa6c5739159ca2a862b50258ad8f61afed4ca94f) including the following fixes:

https://github.com/nemequ/simde/commit/9d29f2b56ff3a93e4f13ab2075fe043ca10b7e7e common: prefer __builtin_isnan over std::isnan 
https://github.com/nemequ/simde/commit/c0af7afcc8290fba98803febb503015c2cf5a4d7 avx2: fix -Wconversion warning when assigning int to int16_t
https://github.com/nemequ/simde/commit/6791c8f891a4153a8736b7b7cdbc6f4bda7e31f8 Switch some C-style casts to use HEDLEY_STATIC_CAST.
https://github.com/nemequ/simde/commit/dbf8394561e699cb11a75a2e26d1efa121f11c79 mmx, sse, sse2: get rid of compound literals
https://github.com/nemequ/simde/commit/7185559a4df6cfb00d4730b9edb7bdb7ac3967b8 common: add -Wc++98-compat-pedantic to list of disabled warnings 
https://github.com/nemequ/simde/commit/f11544f282376cd06f53e694e9d5fd51b640e67d avx: fix aliasing issue in simde_mm_maskload_pd
https://github.com/nemequ/simde/commit/e314ce18967350aec8e899eb2acae59b95e07660 math: fix detection of __builtin_* functions
https://github.com/nemequ/simde/commit/bdf27512d41d583278ae5efee02a636d139fd806 test: disable -Wreserved-id-macro in µnit header. 
https://github.com/nemequ/simde/commit/aa6c5739159ca2a862b50258ad8f61afed4ca94f sse2, sse4.1, avx2: add some additional casts to vector types 

So hopefully the result is less daunting!

**Comment by milot-mirdita** [comment]
I am pretty sure you are running into a genuine infinite loop in the Alignment module currently.

It might the all the instances of code similar to this in StripedSmithWaterman.cpp:
```
cmp = simdi8_movemask(vTemp);
#ifdef AVX2
		if (cmp != 0xffffffff)
#else
		if (cmp != 0xffff)
#endif
```

**Comment by milot-mirdita** [comment]
I still can't manage to compile the code on our servers.

I get many instances of this error (or similar):
```
lib/simde/simde/x86/avx2.h:718:62: error: cannot convert ‘simd_int {aka __vector(4) long long int}’ to ‘__vector(4) long int’ for argument ‘1’ to ‘__vector(4) long int simde_mm256_cmpeq_epi8(__vector(4) long int, __vector(4) long int)’
```

Any idea what's happening?

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.12' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.12)
```

**Comment by mr-c** [comment]
> It might the all the instances of code similar to this in StripedSmithWaterman.cpp:

Ah, I missed those, thanks!

**Comment by mr-c** [comment]
> Any idea what's happening?

Maybe @nemequ does?

Looks like a very old GCC version to me

File: CMakeLists.txt [hunk 1]
```diff
@@ -42,15 +42,15 @@ endif ()
...
-    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -DNEON=1 -DSSE=1 -fsigned-char -mfpu=neon")
+    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -fsigned-char -mfpu=neon")
 elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "^(aarch64.*|AARCH64.*)")
     set(ARM 1)
-    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -DNEON=1 -DSSE=1 -fsigned-char")
+    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -fsigned-char")
 elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "PPC64*|ppc64*|powerpc64*")
     set(PPC64 1)
-    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -DSSE=1 -fsigned-char -std=gnu++0x -mcpu=power8 -mvsx")
+    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -fsigned-char -std=gnu++0x -mcpu=power8 -mvsx")
```
Comment (mr-c) [inline CMakeLists.txt:51]: I think `-mcpu=power8 -mvsx` should be set in CI config, or elsewhere, but not for all PPC64 users

File: CMakeLists.txt [hunk 2]
```diff
@@ -42,15 +42,15 @@ endif ()
 set(MMSEQS_CXX_FLAGS "-std=c++0x")
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "^(arm.*|ARM.*)")
     set(ARM 1)
-    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -DNEON=1 -DSSE=1 -fsigned-char -mfpu=neon")
+    set(MMSEQS_CXX_FLAGS "${MMSEQS_CXX_FLAGS} -fsigned-char -mfpu=neon")
```
Comment (mr-c) [inline CMakeLists.txt:45]: I think `-mfpu=neon` should be set in CI config, or elsewhere, but not for all ARM* users

File: azure-pipelines.yml [hunk 3]
```diff
@@ -87,20 +87,20 @@ jobs:
         submodules: true
       - script: |
           sudo apt-get update
-          sudo apt-get -y install mpi-default-dev mpi-default-bin
+          sudo apt-get -y install mpi-default-dev mpi-default-bin g++-8
```
Comment (mr-c) [inline azure-pipelines.yml:90]: When things calm down, I can remove these upgrades, if desired.

**Comment by milot-mirdita** [comment]
I removed the `goto` in `lca.cpp`. Maybe that helps with the ASAN crash.

**Comment by milot-mirdita** [comment]
SIMDe ARM64 performance seems to be problematic. There is still some minor difference in the PROFILE regression test (AUC: 0.36739*6* vs 0.36739*8*). However the runtime is over 2x slower (14min vs 31min). I think we might want to keep sse2neon at least in the short term until SIMDe can catch. I guess PPCLE will be similar (the work for that was still incomplete though).

HEAD: https://travis-ci.org/github/soedinglab/MMseqs2/jobs/687741954
PR: https://travis-ci.org/github/soedinglab/MMseqs2/jobs/688351766

**Comment by mr-c** [comment]
@milot-mirdita Interesting, thanks for those figures.

What build flags are in effect for https://travis-ci.org/github/soedinglab/MMseqs2/jobs/687741954 ?

**Comment by milot-mirdita** [comment]
It should be nothing more than `-DHAVE_NEON=1` in cmake. Which enables neon and the sse2neon branch.

**Comment by mr-c** [comment]
> It should be nothing more than -DHAVE_NEON=1 in cmake. Which enables neon and the sse2neon branch.

Oh, I meant gcc flags, pardon me.

I think https://github.com/soedinglab/MMseqs2/pull/309/commits/3bde0f7a0cb2e4461e80d6d32d3ec7490e31dfa8 will help a lot !

**Comment by milot-mirdita** [comment]
Only macOS Clang seems to have a problem with this function for some reason:
https://github.com/soedinglab/MMseqs2/blob/master/src/alignment/DistanceCalculator.h#L276

**Comment by milot-mirdita** [comment]
Actually it should be already crashing earlier, since no output is ever printed by `rescorediagonal`.

**Comment by milot-mirdita** [comment]
I checked it's indeed in this function. It doesn't like the simde implementation of `_mm256_loadu_si256` (https://github.com/nemequ/simde/blob/000f9bac0d339d32ec7d06d32f362e381cf4f960/simde/x86/avx.h#L3375).

I think this is a function where the `simd.h` wrapper to generate either native SSE4.1 or AVX2 code was actually pretty useful. We should probably just keep using 128-bit SIMD in this case.

**Comment by milot-mirdita** [comment]
I had to disable the `SIMDE_REQUIRE_CONSTANT` macro completely to compile on AppleClang 11 (`Apple clang version 11.0.0 (clang-1100.0.33.17)`).

**Comment by mr-c** [comment]
@milot-mirdita Are you the author of https://github.com/soedinglab/MMseqs2/blame/master/lib/simd/sse2altivec.h ? Would you mind donating that code to SIMDe?

**Comment by milot-mirdita** [comment]
That's a bit more complicated. We had a contribution to HH-suite to support PPCLE from @dmiller423 and subsequently improvements from @jhuber6. I wanted a different approach similar to `sse2neon` and adapted the code. To re-license the code we might need to ask two more people.

**Comment by milot-mirdita** [comment]
Also the code for that is still incomplete. It was not passing our regression suite yet so there was still something wrong with it.

**Comment by mr-c** [comment]
The ppc64le build now only fails one regression test, the same as arm64 https://travis-ci.org/github/soedinglab/MMseqs2/jobs/688790571#L11062

> PROFILE (Time: 259s)
TEST FAILED
BAD
Expected:  0.367398
Actual:  0.367396

Without this PR the ppc64le build fails 25 regression tests: https://travis-ci.org/github/soedinglab/MMseqs2/jobs/688479773#L10132

**Comment by milot-mirdita** [comment]
I can try to find the regression later with qemu. It’s only a bit painful :D

**Comment by milot-mirdita** [comment]
The remaining issue seems to be that `_mm256_rcp_ps(_mm256_set1_ps(0))` does something different in the non-vectorized loop (inf) and in AVX (some large value).

The issue happens here: https://github.com/soedinglab/MMseqs2/blob/master/src/alignment/PSSMCalculator.cpp#L382 

It's debatable if we are doing something stupid or if SIMDe should allow us to do something stupid ;)

**Comment by milot-mirdita** [comment]
Actually a few lines above something different weird happens:
```
simd_float naa_j = simdi32_i2f(simdi32_set(naa[j]));
```
This results in `2139095040.000000` on x64 and `1` on non-vectorized. The second behavior seems to make more sense. Might be a bug on our side.

**Comment by martin-steinegger** [comment]
@mr-c Thank you for this huge effort to port MMseqs2 to simde.

I think the there are two road block at the moment for this, (1) there is difference in the regression between SSE and the scalar code, (2) the neon code not yet fully supported. 

The first issues is not so easy to fix without having some performance penalty in the SSE branch. The differences betwen SSE and the scalar code originates from the following code:

                        simdf32_store(w_contrib[j] + (a * VECSIZE_INT), simdf32_rcp(res));

The intrinsic `simdf32_rcp` is approximating 1/res but the scalar version is computing the real value. A fix would be to switch from `simdf32_rcp` to `simdf32_div`. However this instruction is much slower and this performance critical code.

The second issue has to be solved in the `simde` abstraction layer. We would require a full support of the instruction for NEON architecture. There is currently a huge speed penalty using `simde` in our regression.

**Comment by mr-c** [comment]
Thank you @martin-steinegger for the acknowledgement

>  there is difference in the regression between SSE and the scalar code

We can insert a branch for this, using `simdf32_rcp` for the AVX+ builds and `simdf32_div` otherwise

>  the neon code not yet fully supported

We are coming up to speed on this right now :-)

**Comment by milot-mirdita** [comment]
I am not sure why Github is not linking to travis (https://travis-ci.org/github/soedinglab/MMseqs2/builds/700050415).

Performance on ARM64 seems much better now, I suspect the remaining performance difference since we only issued 128-bit instructions with the `simd.h` split between 256/128 before and now ARM64 has to do 128-bit instructions from 256-bit instructions in a more round about way.

We could reintroduce the code for that again and see if we get equal performance.

**Comment by mr-c** [comment]
> We could reintroduce the code for that again and see if we get equal performance.

@milot-mirdita Could you push a commit that restores the 128 bit path to this PR? I don't think I understand that part of the code very well.

**Comment by milot-mirdita** [comment]
I tried to here:
https://github.com/milot-mirdita/mmseqs2/tree/simde_128_256_split
https://travis-ci.org/github/milot-mirdita/mmseqs2/builds/700091381

We'll know in a few minutes if that helps.

**Comment by mr-c** [comment]
@milot-mirdita Much faster!

Shall I cherry pick that here? This PR is a bit of a mess now..

**Comment by milot-mirdita** [comment]
I think you can squash everything down to one commit soon. Only a few things remain to do I think. I wrote you in gitter, could you take a look when you have time?

**Comment by milot-mirdita** [comment]
I finished up the integration in #321. Thanks a lot again!

**Pull request closed without merging**